src/Data/Result.php

```diff
@@ -0,0 +1,83 @@
+<?php
+/* Copyright (c) 2017 Richard Klees <richard.klees@concepts-and-training.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\Data;
+
+/**
+ * A result encapsulates a value or an error and simplifies the handling of those.
+ *
+ * To be implemented as immutable object.
+ */
+interface Result {
+	/**
+ 	 * Get to know if the result is ok.
+	 *
+	 * @return bool
+	 */
+	public function isOK();
+
+	/**
+	 * Get the encapsulated value.
+	 *
+	 * @throws Exception    if !isOK, will either throw the contained exception or
+	 *                      a NotOKException if a string is contained as error.
+	 * @return mixed
+	 */
+	public function value();
+
+	/**
+	 * Get to know if the result is an error.
+	 *
+	 * @return bool
+	 */
+	public function isError();
+
+	/**
+	 * Get the encapsulated error.
+	 *
+	 * @throws LogicException   if isOK
+	 * @return Exception|string
+	 */
+	public function error();
+
+	/**
+	 * Get the encapsulated value or the supplied default if result is an error.
+	 *
+	 * @param  default
+	 * @return mixed
+	 */
+	public function valueOr($default);
+
+	/**
+	 * Create a new result where the contained value is modified with $f.
+	 *
+	 * Does nothing if !isOK.
+	 *
+	 * @param	callable $f mixed -> mixed
+	 * @return	Result
+	 */
+	public function map(callable $f);
```

Comment:
If I had a function `$f` which returns `ok("foo")` using it with `map` would yield `ok(ok("foo"))`, while using it with `then` would yield `ok("foo")`. So `then` collapses one layer of Result after the application. `map` can be expressed in terms of `then`, but not the other way round. `map` belongs to the functor abstraction, `then` to monad.